### PR TITLE
WIP: Create a new kubectl command to attach debug container to a running pod

### LINKED ISSUE #45922: FR: New kubectl command `kubectl debug`
**Is this a BUG REPORT or FEATURE REQUEST?** (choose one): FEATURE REQUEST

SIG Node is working on new functionality (feature: kubernetes/features#277 , proposal: #35584) to execute a "Debug Container" in the context of a running pod for the purposes of troubleshooting. This issue is to discuss changes needed to kubectl to surface this feature.

We're aiming for an alpha release in 1.7 that includes basic functionality. As written, the proposal in #35584 calls for a new command, `kubectl debug`, that resembles `kubectl exec`. An example run is:

```
% kubectl debug -it --image alpine $POD -c $DEBUG_CONTAINER_NAME -- sh
/ # ps x
PID   USER     TIME   COMMAND
    1 root       0:00 /pause
   13 root       0:00 /app
   26 root       0:00 sh
   32 root       0:00 ps x
/ #
```

Ideally we could have reasonable defaults so that the minimum command would be something like:
```
% kubectl debug -it
Defaulting container name to debug.
/ # 
```

One departure from `kubectl exec` is that `kubectl debug` supports reattaching to a running Debug Container with a command like `kubectl debug --reattach $POD $CONTAINER_NAME`

Options for the command would resemble:
```
Execute a Debug Container in a Pod.

Options:
  -c, --container='': Name of container to create. If omitted, the default "debug" will be used.
  -i, --stdin=false: Pass stdin to the container
  -m, --image='': Container Image to use when creating debug container. If omitted, the cluster default is used.
  -t, --tty=false: Stdin is a TTY

Usage:
  kubectl debug POD [-m IMAGE ] [-c CONTAINER] -- COMMAND [args...] [options]
```

Since this is an alpha feature it might be nice to be able to hide it in kubectl unless alpha features are enabled. It will return an error on clusters with alpha features disabled (the default).

/cc @pwittrock

### PR DESCRIPTION
**Pull request by verb**
**What this PR does / why we need it**: This is the client command to attach a debug container in a running pod, as proposed in kubernetes/community#649

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #45922

**Special notes for your reviewer**: This PR depends on kubectl feature gates #46151

**Release note**:

```release-note
NONE
```

**Dependencies:**
- [ ] #46151 Alpha feature gating for `kubectl`

**WIP:**
- [ ] Need decision from api-machinery on API endpoint

### CODE REVIEW

File: pkg/kubectl/cmd/cmd.go [hunk 1]
```diff
@@ -252,6 +255,12 @@ var (
 
 // NewKubectlCommand creates the `kubectl` command and its nested children.
 func NewKubectlCommand(f cmdutil.Factory, in io.Reader, out, err io.Writer) *cobra.Command {
+	if f := os.Getenv("KUBECTL_FEATURE_GATES"); f != "" {
+		if e := utilfeature.DefaultFeatureGate.Set(f); e != nil {
+			fmt.Fprintf(err, "Invalid KUBECTL_FEATURE_GATES %q. Available features:\n%s\n", f, strings.Join(utilfeature.DefaultFeatureGate.KnownFeatures(), "\n"))
```
Comment (alexandercampbell) [inline pkg/kubectl/cmd/cmd.go]: Maybe do `strings.Join` with `"\n\t"` instead of just `"\n"`?

File: pkg/kubectl/cmd/templates/command_groups.go [hunk 2]
```diff
@@ -35,6 +35,20 @@ func (g CommandGroups) Add(c *cobra.Command) {
...
+	for i, group := range g {
+		gc := make([]*cobra.Command, 0, len(group.Commands))
+		for _, c := range group.Commands {
+			if c != nil {
+				gc = append(gc, c)
+			}
+		}
+		g[i].Commands = gc
+	}
+}
```
Comment (alexandercampbell) [inline pkg/kubectl/cmd/templates/command_groups.go]: This means that `kubectl help debug` will fail to print the documentation if the feature is disabled by the environment.
I assume this is what you want, but what if it instead printed information about how to enable the command?

File: pkg/kubectl/cmd/debug.go [hunk 3]
```diff
@@ -0,0 +1,231 @@
...
+			Out: cmdOut,
+			Err: cmdErr,
+		},
+
+		Debugger: &DefaultRemoteDebugger{},
+	}
+	cmd := &cobra.Command{
+		Use:     "debug POD [-c CONTAINER] -- COMMAND [args...]",
+		Short:   i18n.T("Run a debug container in a pod"),
+		Long:    "Run a new container inside an existing pod.",
```
Comment (alexandercampbell) [inline pkg/kubectl/cmd/debug.go]: I feel like the Long help message should also include the word "debug".

File: pkg/kubectl/cmd/debug.go [hunk 4]
```diff
@@ -0,0 +1,231 @@
...
+// Validate checks that the provided debug options are specified.
+func (p *DebugOptions) Validate() error {
+	if len(p.PodName) == 0 {
+		return fmt.Errorf("pod name must be specified")
+	}
+	if p.Out == nil || p.Err == nil {
+		return fmt.Errorf("both output and error output must be provided")
+	}
+	if p.Debugger == nil || p.PodClient == nil || p.Config == nil {
+		return fmt.Errorf("client, client config, and debugutor must be provided")
```
Comment (alexandercampbell) [inline pkg/kubectl/cmd/debug.go]: What is a debugutor?

File: pkg/kubectl/cmd/debug.go [hunk 5]
```diff
@@ -0,0 +1,231 @@
...
+		}, api.ParameterCodec)
+
+		return p.Debugger.Debug("POST", req.URL(), p.Config, p.In, p.Out, p.Err, t.Raw, sizeQueue)
+	}
+
+	if err := t.Safe(fn); err != nil {
+		return err
+	}
+
+	return nil
```
Comment (alexandercampbell) [inline pkg/kubectl/cmd/debug.go]: Just do

```go
return t.Safe(fn)
```

File: pkg/kubectl/cmd/debug.go [hunk 6]
```diff
@@ -0,0 +1,231 @@
...
+)
+
+const (
+	debugUsageStr             = "expected 'debug POD_NAME [--] [COMMAND [ARG1 ... ARGN]]'.\nPOD_NAME is a required argument for the debug command.\nCOMMAND may be required if the container does not provide an ENTRYPOINT"
+	debugDefaultContainerName = "debug"
+	debugDefaultImageName     = "debian"
+)
+
+func NewCmdDebug(f cmdutil.Factory, cmdIn io.Reader, cmdOut, cmdErr io.Writer) *cobra.Command {
+	if !utilfeature.DefaultFeatureGate.Enabled(features.DebugContainers) {
```
Comment (pwittrock) [inline pkg/kubectl/cmd/debug.go]: Having kubectl depend on Kubernetes repo is an anti pattern.  Don't do this.  It either needs to go in one of the staging repos (client-go, etc) or be copied into the kubectl packages

File: pkg/kubectl/cmd/debug.go [hunk 7]
```diff
@@ -0,0 +1,231 @@
+/*
+Copyright 2014 The Kubernetes Authors.
```
Comment (pwittrock) [inline pkg/kubectl/cmd/debug.go]: copyright

File: pkg/kubectl/cmd/debug.go [hunk 8]
```diff
@@ -0,0 +1,231 @@
...
+	"net/url"
+
+	"github.com/spf13/cobra"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	remotecommandconsts "k8s.io/apimachinery/pkg/util/remotecommand"
+	utilfeature "k8s.io/apiserver/pkg/util/feature"
+	restclient "k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/remotecommand"
+	"k8s.io/kubernetes/pkg/api"
```
Comment (pwittrock) [inline pkg/kubectl/cmd/debug.go]: New imports from k8s.io/kubernetes are forbidden

File: pkg/kubectl/cmd/debug.go [hunk 9]
```diff
@@ -0,0 +1,231 @@
...
+	coreclient "k8s.io/kubernetes/pkg/client/clientset_generated/internalclientset/typed/core/internalversion"
+	"k8s.io/kubernetes/pkg/features"
+	"k8s.io/kubernetes/pkg/kubectl/cmd/templates"
+	cmdutil "k8s.io/kubernetes/pkg/kubectl/cmd/util"
+	"k8s.io/kubernetes/pkg/util/i18n"
```
Comment (pwittrock) [inline pkg/kubectl/cmd/debug.go]: this one too

File: pkg/kubectl/cmd/debug.go (hunk 6 above)
Comment (verb) [inline pkg/kubectl/cmd/debug.go]: @sttts mentioned wanting to use a separate feature gate in https://github.com/kubernetes/kubernetes/pull/46151#issuecomment-303032518, I imagine this must be why. Can kubectl depend on apiserver?

I have little preference on how to hide alpha features, so I need guidance from those who do. I mention some trade-offs here: https://github.com/kubernetes/kubernetes/issues/45922#issuecomment-302895410

**Comment by pwittrock** [comment]
Hm.  I don't want to block this functionality on kubectl deps issues which are hard to address, but don't want to create a bunch of additional work to pull them out.

If we accept the deps for this PR, will you remove the deps after code freeze?

**Comment by verb** [comment]
@pwittrock yesterday there arose some disagreement in api-machinery about this API change, so I'm happy to take a little more time to get deps/alpha gate right and miss the 1.7 deadline. I'm writing up the suggested API change and would appreciate your opinion on it.

**Comment by pwittrock** [comment]
@verb great.  that takes the pressure off

File: pkg/kubectl/cmd/debug.go (hunk 4 above)
Comment (verb) [inline pkg/kubectl/cmd/debug.go]: a find/replace fail... 

Fixed, thanks.

File: pkg/kubectl/cmd/debug.go (hunk 5 above)
Comment (verb) [inline pkg/kubectl/cmd/debug.go]: Fixed, thanks.

File: pkg/kubectl/cmd/templates/command_groups.go (hunk 2 above)
Comment (verb) [inline pkg/kubectl/cmd/templates/command_groups.go]: Yup, that's what I intended here, but what you describe could definitely be a better user experience. I'll try to figure out something for this in #46151

**Comment by alexandercampbell** [comment]
@verb is this ready for re-review?

**Comment by verb** [comment]
@alexandercampbell Not quite yet, I need resolution from api-machinery about naming the api endpoint.

**Comment by alexandercampbell** [comment]
@verb close for now and re-open when the apimachinery question is resolved?

**Comment by verb** [comment]
@alexandercampbell SGTM. Thanks for your patience.

**Comment by verb** [comment]
@r2d4 FYI, this is the example `kubectl debug` PR. It doesn't have the alpha integration, but you can see where it would have gone here: https://github.com/kubernetes/kubernetes/blob/master/pkg/kubectl/cmd/alpha.go#L39

**Pull request closed without merging**